ssl/quic/quic_fc.c

```diff
@@ -0,0 +1,329 @@
+{
+    int r = txfc->has_become_blocked;
+
+    if (clear)
+        txfc->has_become_blocked = 0;
+
+    return r;
+}
+
+uint64_t ossl_quic_txfc_get_cwm(QUIC_TXFC *txfc)
+{
+    return txfc->cwm;
+}
+
+uint64_t ossl_quic_txfc_get_swm(QUIC_TXFC *txfc)
+{
+    return txfc->swm;
+}
+
+/*
+ * RX Flow Controller (RXFC)
+ * =========================
+ */
+
+int ossl_quic_rxfc_init(QUIC_RXFC *rxfc, int is_stream,
+                        uint64_t initial_window_size,
+                        uint64_t max_window_size,
+                        OSSL_TIME (*now)(void *now_arg),
+                        void *now_arg)
+{
+    rxfc->swm               = 0;
+    rxfc->cwm               = initial_window_size;
+    rxfc->rwm               = 0;
+    rxfc->esrwm             = 0;
+    rxfc->hwm               = 0;
+    rxfc->cur_window_size   = initial_window_size;
+    rxfc->max_window_size   = max_window_size;
+    rxfc->is_stream         = is_stream;
+    rxfc->error_code        = 0;
+    rxfc->has_cwm_changed   = 0;
+    rxfc->epoch_start       = ossl_time_zero();
+    rxfc->now               = now;
+    rxfc->now_arg           = now_arg;
+    rxfc->is_fin            = 0;
+    return 1;
+}
+
+static void rxfc_start_epoch(QUIC_RXFC *rxfc)
+{
+    rxfc->epoch_start   = rxfc->now(rxfc->now_arg);
+    rxfc->esrwm         = rxfc->rwm;
+}
+
+static int on_rx_controlled_bytes(QUIC_RXFC *rxfc, uint64_t num_bytes)
+{
+    int ok = 1;
+    uint64_t credit = rxfc->cwm - rxfc->swm;
+
+    if (num_bytes > credit) {
+        ok = 0;
+        num_bytes = credit;
+        rxfc->error_code = QUIC_ERR_FLOW_CONTROL_ERROR;
+    }
+
+    rxfc->swm += num_bytes;
+    return ok;
+}
+
+int ossl_quic_rxfc_on_rx_stream_frame(QUIC_RXFC *rxfc, QUIC_RXFC *conn_rxfc,
+                                      uint64_t end, int is_fin)
+{
+    uint64_t delta;
+
+    if (!rxfc->is_stream || conn_rxfc == NULL || conn_rxfc->is_stream)
+        return 0;
+
+    if (rxfc->is_fin && ((is_fin && rxfc->hwm != end) || end > rxfc->hwm)) {
+        /* Stream size cannot change after the stream is finished */
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
+
+    if (is_fin)
+        rxfc->is_fin = 1;
+
+    if (end > rxfc->hwm) {
+        delta = end - rxfc->hwm;
+        rxfc->hwm = end;
+
+        on_rx_controlled_bytes(rxfc, delta);        /* result ignored */
+        on_rx_controlled_bytes(conn_rxfc, delta);   /* result ignored */
+    } else if (end < rxfc->hwm && is_fin) {
+        rxfc->error_code = QUIC_ERR_FINAL_SIZE_ERROR;
+        return 1; /* not a caller error */
+    }
+
+    return 1;
+}
+
+/* threshold = 3/4 */
+#define WINDOW_THRESHOLD_NUM 3
+#define WINDOW_THRESHOLD_DEN 4
+
+static int rxfc_cwm_bump_desired(QUIC_RXFC *rxfc)
+{
+    uint64_t window_rem = rxfc->cwm - rxfc->rwm;
+    uint64_t threshold
+        = (rxfc->cur_window_size * WINDOW_THRESHOLD_NUM) / WINDOW_THRESHOLD_DEN;
```

Comment:
I think this should use the `safe_muldiv_` function from `internal/safe_math.h`.  It avoids the possibility of overflow.